design/permissions.md

```diff
@@ -0,0 +1,198 @@
+Permissions API
+=================
+
+To be a responsible member of the Social MR ecosystem, the MRE SDK has to prioritize the safety of its residents over
+the power of this SDK. As such, I propose the addition of a permissions API, so MREs can have more "invasive" features
+without compromising user safety and agency.
+
+
+Permission scope
+------------------
+
+When a user approves or denies permission to an MRE, that approval/denial is scoped to the unique combination of
+protocol, hostname, and port of the MRE URL, known collectively as the "origin". So when `ws://mres.altvr.com/wearahat`
+requests `user-interaction`, that permission is associated with `ws://mres.altvr.com`, and all user decisions apply
+equally to new connections to the same server. These user decisions may or may not be persisted indefinitely by the host app.
+
+
```

Comment:
I think the host app should decide the scope here.
I'd also recommend rights to be managed by AppId over URL, if AppId is available

Reply:
Could you explain your reasoning for hosts making the scope determination? I can see the case for AppId over URL origin, because AppIds can uniquely identify an app and not just a server. I disagree, and think server trust is all that's required since there won't be origins with apps from multiple authors, and it's the author that the user is ultimately trusting.

---
In reply to: [441849523](https://github.com/microsoft/mixed-reality-extension-sdk/pull/606#discussion_r441849523) [](ancestors = 441849523)

Reply:
I should have worded that differently. I wanted to say that I don't think we should assume that scope goes beyond the single app, and the host app should ask the user if they would want to permit all apps, all apps by same developer, or a single app.

I, as a user, wouldn't want the assumption made about other apps for me. Just because I want to grant one app a specific permission, doesn't mean that all other apps that are deployed to the same root url should have the same permissions.

Reply:
So how do you define a single app, the full URL? Does it include query arguments? The more specific we get, the more nag screens the users are gonna be presented with. I think origin strikes a good balance.

Reply:
Yeah, a unique app to me is the full URL but excluding query arguments. 
I think the best analogy is phone apps: If I install an app on my phone and I grant a permission, it shouldn't automatically grant permission to all other apps by that publisher. But if I launch the app with a different deep link, the permissions given to the app before still apply.

Agree that there will be more nag screens. To me that's up to the host app for how to manage. I imagine "yes/no once", "yes/no for this MRE always", "yes/no for all MREs always" would be required. It could be optional to add "yes/no for all MREs on this domain" (or "by this publisher" once MRE App registry is done), but I see that as lower priority